packages/apollo-server-codegen/src/IR/index.ts

```diff
@@ -0,0 +1,75 @@
+import { DocumentNode } from "graphql";
+import { Description } from "./Descriptions";
+import { EnumDefinition } from "./Enums";
+import {
+  FieldDefinition,
+  ObjectDefinition,
+  TypelessObjectDefinition
+} from "./Objects";
+import { ArgumentDefinition, ResolverDefinition } from "./Resolvers";
+import { ScalarDefinition } from "./Scalars";
+import { CompoundType, ListType, NamedType, NonNullType } from "./Types";
+
+export {
+  ResolverDefinition,
+  ArgumentDefinition,
+  Description,
+  FieldDefinition,
+  ObjectDefinition,
+  CompoundType,
+  ListType,
+  NamedType,
+  NonNullType,
+  EnumDefinition,
+  ScalarDefinition
+};
+
+export function sdlToIR(
+  sdl: DocumentNode
+): [ObjectDefinition[], EnumDefinition[], ScalarDefinition[]] {
+  const objectDefinitions: TypelessObjectDefinition[] = [];
+  const enumDefinitions = [];
+  const scalarDefinitions = [];
+  for (const definition of sdl.definitions) {
```

Comment:
Personal preference, but this seems like a nice use case for `graphql`'s `visit()` fn. Just recommending it in case you're unfamiliar.

Here's one example, but it's littered throughout federation / AS code. It's great for traversing SDL (and modifying / returning new SDL if you choose!).

https://github.com/apollographql/apollo-server/blob/42ab6b86b7fb3adcdb5f3fd1036cfa2b7a524803/packages/apollo-federation/src/composition/normalize.ts#L132

Reply:
Going to keep it as is for now, as to me this more explicitly shows that we're only interested in top-level definitions (and that we're treating certian classes of definitions (type definitions, type extensions, interface definition, etc.) exactly the same.